### PR TITLE
[T6811]bad_connector_woocommerce:Export Stock update BAD-SIP.

### PR DESCRIPTION
**Pull request by bizzappdevteam**
- Import the stock Manage settings from woocommerce which will be stored in woo.settings so it will be imported at the time of 'sync Metadata'
- if stock_manage setting is true at product level then its product_type/detailed_type value will be 'storable product'
- Export the updated stock quantity and its button is located at the backend level and at the product level.
- Change the backend view for the import tab.

### CODE REVIEW

File: bad_connector_woocommerce/models/product/common.py [hunk 1]
```diff
@@ -1,12 +1,19 @@
...
 
 from odoo import fields, models
 
 from odoo.addons.component.core import Component
+from odoo.addons.component_event import skip_if
 
 _logger = logging.getLogger(__name__)
 
 
+def chunks(items, length):
```
Comment (bizzappdev) [inline bad_connector_woocommerce/models/product/common.py]: move this to utils

File: bad_connector_woocommerce/models/product/common.py [hunk 2]
```diff
@@ -28,6 +40,8 @@ class WooProductProduct(models.Model):
 
     _rec_name = "name"
 
+    RECOMPUTE_QTY_STEP = 1000  # products at a time
```
Comment (bizzappdev) [inline bad_connector_woocommerce/models/product/common.py]: make it configurable from backend

File: bad_connector_woocommerce/models/product/common.py [hunk 3]
```diff
@@ -17,6 +24,11 @@ class ProductProduct(models.Model):
         copy=False,
     )
 
+    def update_stock_qty(self):
+        """Import Product Attribute Value"""
```
Comment (bizzappdev) [inline bad_connector_woocommerce/models/product/common.py]: proper doc string

File: bad_connector_woocommerce/models/product/common.py [hunk 4]
```diff
@@ -85,6 +99,82 @@ class WooProductProduct(models.Model):
...
+        It groups the products by backend to avoid to read the backend
+        informations for each product.
+        """
+        # group products by backend
+        backends = defaultdict(set)
+        for product in self:
+            backends[product.backend_id].add(product.id)
+        for backend, product_ids in backends.items():
+            self._recompute_woo_qty_backend(backend, self.browse(product_ids))
+        return True
```
Comment (bizzappdev) [inline bad_connector_woocommerce/models/product/common.py]:
```suggestion
        backends = defaultdict(self.env['woo.product.product'])
        for woo_product in self:
            backends[woo_product.backend_id] |= woo_product
        for backend, woo_products in backends.items():
            self._recompute_woo_qty_backend(backend, woo_products)
        return True
```

File: bad_connector_woocommerce/models/product/common.py [hunk 5]
```diff
@@ -85,6 +99,82 @@ class WooProductProduct(models.Model):
...
+        product_fields = ["woo_product_qty", stock_field]
+        if read_fields:
+            product_fields += read_fields
+
+        self_with_location = self.with_context(location=location.id)
+        for chunk_ids in chunks(products.ids, self.RECOMPUTE_QTY_STEP):
+            records = self_with_location.browse(chunk_ids)
+            for product in records.read(fields=product_fields):
+                new_qty = self._woo_qty(product, backend, location, stock_field)
+                if new_qty != product["woo_product_qty"]:
```
Comment (bizzappdev) [inline bad_connector_woocommerce/models/product/common.py]: use if not continue

File: bad_connector_woocommerce/models/product/importer.py [hunk 6]
```diff
@@ -344,3 +358,17 @@ def _after_import(self, binding, **kwargs):
...
         return result
+
+
+class ProductInventoryExporter(Component):
+    _name = "woo.product.product.exporter"
+    _inherit = "woo.exporter"
+    _apply_on = ["woo.product.product"]
+    _usage = "product.inventory.exporter"
+
+    def run(self, binding, fields):
```
Comment (bizzappdev) [inline bad_connector_woocommerce/models/product/importer.py]: have the exact signature from the exporter

File: bad_connector_woocommerce/models/woo_backend/common.py [hunk 7]
```diff
@@ -440,3 +480,25 @@ def cron_import_metadata(self, domain=None):
         """Cron for sync_metadata"""
         backend_ids = self.search(domain or [])
         backend_ids.sync_metadata()
+
+    def _domain_for_update_product_stock_qty(self):
+        """Domain to search WooCommerce product"""
+        return [
+            ("backend_id", "in", self.ids),
+            ("type", "!=", "service"),
```
Comment (bizzappdev) [inline bad_connector_woocommerce/models/woo_backend/common.py]: better to add stock and IMO it is not type by detailed type

File: bad_connector_woocommerce/models/product/importer.py [hunk 8]
```diff
@@ -323,6 +323,20 @@ def woo_product_attribute_value_ids(self, record):
                 attribute_value_ids.append(attribute_value.id)
         return {"woo_product_attribute_value_ids": [(6, 0, attribute_value_ids)]}
 
+    @mapping
+    def stock_management(self, record):
+        """Mapping for Stock Management"""
+        return {"stock_management": record.get("manage_stock")}
```
Comment (bizzappdev) [inline bad_connector_woocommerce/models/product/importer.py:332]: if manage stock is true then also set proper product type

**Review by bizzappdev — CHANGES_REQUESTED** [review]

File: bad_connector_woocommerce/models/product/common.py [hunk 9]
```diff
@@ -85,6 +97,83 @@ class WooProductProduct(models.Model):
...
+        """
+        if backend.product_stock_field_id:
+            stock_field = backend.product_stock_field_id.name
+        else:
+            stock_field = "virtual_available"
+
+        location = self.env["stock.location"]
+        if self.env.context.get("location"):
+            location = location.browse(self.env.context["location"])
+        else:
```
Comment (bizzappdev) [inline bad_connector_woocommerce/models/product/common.py]: remove this condition and have it from warehouse always

File: bad_connector_woocommerce/models/product/common.py [hunk 10]
```diff
@@ -85,6 +97,83 @@ class WooProductProduct(models.Model):
...
+            location = backend.warehouse_id.lot_stock_id
+        product_fields = ["woo_product_qty", stock_field]
+        if read_fields:
+            product_fields += read_fields
+
+        self_with_location = self.with_context(location=location.id)
```
Comment (bizzappdev) [inline bad_connector_woocommerce/models/product/common.py]: sudo

File: bad_connector_woocommerce/models/product/common.py [hunk 11]
```diff
@@ -85,6 +97,83 @@ class WooProductProduct(models.Model):
...
+        for chunk_ids in utils.chunks(products.ids, backend.recompute_qty_step):
+            records = self_with_location.browse(chunk_ids)
+            for product in records.read(fields=product_fields):
```
Comment (bizzappdev) [inline bad_connector_woocommerce/models/product/common.py]: use browse instead

**Review by bizzappdev — CHANGES_REQUESTED** [review]

File: bad_connector_woocommerce/models/product/common.py [hunk 12]
```diff
@@ -85,6 +107,77 @@ class WooProductProduct(models.Model):
...
+        if read_fields:
+            product_fields += read_fields
+        self_with_location = self.sudo().with_context(location=location.id)
+        for chunk_ids in utils.chunks(products.ids, backend.recompute_qty_step):
+            records = self_with_location.browse(chunk_ids)
+            for product in records:
+                new_qty = self._woo_qty(product, backend, location, stock_field)
+                if not new_qty != product["woo_product_qty"]:
+                    continue
+                self.browse(product["id"]).woo_product_qty = new_qty
```
Comment (bizzappdev) [inline bad_connector_woocommerce/models/product/common.py]:
```suggestion
                product.woo_product_qty = new_qty
```

**Review by bizzappdev — CHANGES_REQUESTED** [review]

File: bad_connector_woocommerce/migrations/16.0.3.0.1/post-migration.py [hunk 13]
```diff
@@ -0,0 +1,11 @@
+def migrate(cr, version):
```
Comment (bizzappdev) [inline bad_connector_woocommerce/migrations/16.0.3.0.1/post-migration.py]: no need for migration script make field required only from view

**Comment by bizzappdevteam** [comment]
This PR is merged in #52.

**Pull request closed without merging**